h/api/storage.py

```diff
@@ -61,12 +74,77 @@ def create_annotation(request, data):
     :returns: the created annotation
     :rtype: dict
     """
-    annotation = elastic.Annotation(data)
-
-    # FIXME: this should happen when indexing, not storing.
-    _prepare(request, annotation)
+    if not request.feature('postgres_write'):
+        return _legacy_create_annotation_in_elasticsearch(request, data)
+
+    document_uri_dicts = data['document']['document_uri_dicts']
+    document_meta_dicts = data['document']['document_meta_dicts']
+    del data['document']
+
+    # Replies must have the same group as their parent.
+    if data['references']:
+        top_level_annotation_id = data['references'][0]
+        top_level_annotation = fetch_annotation(request,
+                                                top_level_annotation_id)
+        if top_level_annotation:
+            data['groupid'] = top_level_annotation.groupid
+        else:
+            raise schemas.ValidationError(
+                'references.0: ' +
+                _('Annotation {annotation_id} does not exist').format(
+                    annotation_id=top_level_annotation_id)
+            )
+
+    # The user must have permission to create an annotation in the group
+    # they've asked to create one in.
+    if data['groupid'] == '__world__':
+        group_principal = security.Everyone
```

Comment:
You still need to check that the user is a member of the group to which you're about to write the annotations.